framework/src/media/FileInputDataSource.cpp

```diff
@@ -96,46 +96,50 @@ bool FileInputDataSource::isPrepare()
 
 size_t FileInputDataSource::read(unsigned char *buf, size_t size)
 {
-	size_t ret;
+	size_t rlen = 0;
 	size_t readSize = size;
 	std::shared_ptr<Decoder> decoder = getDecoder();
 	if (decoder) {
-		size_t frames = 0;
-		/* TODO Currently it just return frames which decoded only ONCE if available */
-		ret = getDecodeFrames(buf, &frames);
-		medvdbg("decode frame : %d\n", frames);
-		if (ret > 0) {
-			return ret;
+		size_t frames = size;
+		rlen = getDecodeFrames(buf, &frames);
+		medvdbg("decode frame : %d/%d\n", rlen, size);
+		if (rlen == size) {
+			return rlen;
 		}
 
+		/* Set buf offset */
+		buf += rlen;
+		/* Calculate available space in 'buf' */
+		readSize -= rlen;
+
 		if (readSize > decoder->getAvailSpace()) {
 			readSize = decoder->getAvailSpace();
 		}
 	}
 
-	ret = fread(buf, sizeof(unsigned char), readSize, mFp);
 	/* If file position reaches end of file, return 0 */
 	if (feof(mFp)) {
 		medvdbg("eof!! stop!\n");
-		return 0;
+		return rlen;
 	}
 
+	size_t ret = fread(buf, sizeof(unsigned char), readSize, mFp);
 	medvdbg("read size : %d\n", ret);
 	if (ret == 0) {
-		return EOF;
+		return rlen;
```

Comment:
please check playback() in MediaPlayerImpl.
0 means finished, negative means error

Reply:
``` C
size_t read(unsigned char* buf, size_t size);
```
Return value type is size_t, how can we return a negative value ?
In addition, in payback(), is it possible to reach else { } ?

``` C
void MediaPlayerImpl::playback()
{
	size_t num_read = mInputDataSource->read(mBuffer,(int)mBufSize);
	if (num_read > 0) {
		...
	} else if (num_read == 0) {
		...
	} else {  // num_read < 0 ?
		... 
	}
}
```

Reply:
then size_t should be modified to check condition. 
some value for error case, some value for playback ended, some value for playing

Reply:
Do you mean change `size_t` to `ssize_t` like this? and how about write method?
``` C
ssize_t InputDataSource::read(unsigned char* buf, size_t size);
??? OutputDataSource::write(unsigned char* buf, size_t size);
```
Should these be modified in this PR?

Reply:
If we can distinguish each result, we can accept whatever we change file type. =)
I think ssize_t is more comfortable than size_t for now. like a write & read in file system.